Source/FluentSpreadsheets/Sheets/VisitorImplementations/SheetSegmentVisitor.cs

```diff
@@ -0,0 +1,174 @@
+    IPrototypeHeaderSegmentVisitor<THeaderData>
+{
+    private readonly THeaderData _headerData;
+    private readonly IReadOnlyCollection<TRowData> _rowData;
+    private readonly TFooterData _footerData;
+    private SheetSegmentModel? _segment;
+
+    public SheetSegmentVisitor(
+        THeaderData headerData,
+        IReadOnlyCollection<TRowData> rowData,
+        TFooterData footerData)
+    {
+        _headerData = headerData;
+        _rowData = rowData;
+        _footerData = footerData;
+    }
+
+    public SheetSegmentModel BuildSheetSegment(ISheetSegment builder)
+    {
+        builder.Accept(this);
+
+        _segment ??= Visit(builder);
+
+        if (builder is IPrototypeSegmentHeaderCustomizer<THeaderData> headerCustomizer && _segment.HeaderComponent is not null)
+        {
+            var width = _segment.HeaderComponent.Size.Width;
+            var customizedHeader = headerCustomizer.CustomizeHeader(_segment.HeaderComponent, _headerData);
+
+            if (!customizedHeader.Size.Width.Equals(width))
+            {
+                const string message = "The header customizer must not change the width of the header component.";
+                throw new InvalidStructureException(message);
+            }
+            
+            _segment.HeaderComponent = customizedHeader;
+        }
+
+        var segment = _segment;
+        _segment = null;
+
+        return segment;
+    }
+
+    public void Visit<TDestination>(IPrototypeHeaderSegment<THeaderData, TDestination> builder)
+    {
+        TDestination[] data = builder.Select(_headerData).ToArray();
+
+        var headerVisitor = new PrototypeHeaderSegmentVisitor<TDestination>(data);
+        var rowVisitor = new PrototypeRowSegmentVisitor<TDestination, TRowData>(data, _rowData);
+        var footerVisitor = new PrototypeFooterSegmentVisitor<TDestination, TFooterData>(data, _footerData);
+
+        builder.Accept(headerVisitor);
+        builder.Accept(rowVisitor);
+        builder.Accept(footerVisitor);
+
+        IList<IComponent> headerComponents = headerVisitor.Components;
+        IList<IList<IComponent>> rowComponents = rowVisitor.Components;
+        IList<IComponent> footerComponents = footerVisitor.Components;
+
+        ScalePrototypedComponents(data, headerComponents, rowComponents, footerComponents);
+
+        IComponent? header = null;
+        IList<IComponent> rows = Array.Empty<IComponent>();
+        IComponent? footer = null;
+
+        if (headerComponents.Count is not 0)
+            header = new HStackComponent(headerComponents);
+
+        if (rowComponents.Count is not 0)
+        {
+            var convertedRows = ConvertRowsFromSegmentGroupingToRowGrouping(rowComponents, _rowData.Count);
+
+            rows = convertedRows
+                .Select(x => (IComponent)new HStackComponent(x))
+                .ToArray();
+        }
+
+        if (footerComponents.Count is not 0)
```

Comment:
Maybe make `.Count is not 0` an extension?

Reply:
Pattern matching is kinda reads and looks better that `footerComponents.ItNotEmpty()`

And we'll have `footerComponents is not []` in C# 11, so I would vote against these kinds of extension methods

Reply:
For me, `.Count is not 0` just looks odd, especially giving the fact that you need to write it correctly every time. But after all, its your library